src/pymor/algorithms/system.py

```diff
@@ -0,0 +1,227 @@
+
+    @match_class(BlockOperator)
+    def action_BlockOperator(self, op):
+        range_bases, source_bases, system_in_range, system_in_source, products = \
+            self.range_bases, self.source_bases, self.system_in_range, self.system_in_source, self.products
+        if not system_in_source:
+            raise NotImplementedError
+        if not system_in_range:
+            raise NotImplementedError
+
+        def project_block(i, j, block):
+            return project(block, range_bases[i], source_bases[j], products[i])
+
+        projected_blocks = np.array([project_block(i, j, block) if not isinstance(block, ZeroOperator) else None
+                                     for (i, j), block in np.ndenumerate(op.blocks)])
+        projected_blocks.shape = op.blocks.shape
+        return BlockOperator(blocks=projected_blocks)
+
+    @match_class(BlockRowOperator)
+    def action_BlockRowOperator(self, op):
+        range_bases, source_bases, system_in_range, system_in_source, products = \
+            self.range_bases, self.source_bases, self.system_in_range, self.system_in_source, self.products
+        assert not system_in_range
+        assert system_in_source  # Other case is handled by action_no_system
+
+        def project_block(j, block):
+            return project(block, range_bases, source_bases[j], products)
+
+        projected_blocks = [project_block(j, o) if not isinstance(o, ZeroOperator) else None
+                            for j, o in enumerate(op.blocks)]
+        return BlockRowOperator(projected_blocks)
+
+    @match_class(BlockColumnOperator)
+    def action_BlockColumnOperator(self, op):
+        range_bases, source_bases, system_in_range, system_in_source, products = \
+            self.range_bases, self.source_bases, self.system_in_range, self.system_in_source, self.products
+        assert not system_in_source
+        assert system_in_range  # Other case is handled by action_no_system
+
+        def project_block(i, block):
+            return project(block, range_bases[i], source_bases, product=products[i])
+
+        projected_blocks = [project_block(i, o) if not isinstance(o, ZeroOperator) else None
+                            for i, o in enumerate(op.blocks)]
+        return BlockColumnOperator(projected_blocks)
+
+    @match_class(VectorArrayOperator)
+    def action_VectorArrayOperator(self, op):
+        range_bases, source_bases, system_in_range, system_in_source, products = \
+            self.range_bases, self.source_bases, self.system_in_range, self.system_in_source, self.products
+        if not op.transposed:
+            assert not system_in_source
+            if source_bases is not None:
+                raise NotImplementedError
+            assert system_in_range  # Other case is handled by action_no_system
+
+            def inner(U, V, product=None):
+                return U.dot(V) if product is None else product.apply2(U, V)
+
```

Comment:
would it make sense to allow that either `U` or `V` is `None` here?
projection of a block system only with regard to the 2nd variable:
```python
op = BlockOperator([[A, B], [C, D]])
rhs_blocks = (E, F)
rhs = VectorOperator(op.source.make_array(rhs_blocks))
pop = project_system(op, [None, V], [None, V])
prhs = project_system(rhs, [None, V], None)
# result
# pop.blocks = [[A, B V], [V.T C, V.T D V]]
# prhs.blocks = [E, V.T F]
```

Reply:
Yeah, that should be made possible ..